frontend/viewer/src/lib/services/event-bus.ts

```diff
@@ -1,9 +1,33 @@
-﻿import type {CloseReason} from '../generated-signalr-client/TypedSignalR.Client/Lexbox.ClientServer.Hubs';
-import type {IEntry} from '$lib/dotnet-types';
+﻿import {type CloseReason} from '../generated-signalr-client/TypedSignalR.Client/Lexbox.ClientServer.Hubs';
+import {DotnetService, type IEntry} from '$lib/dotnet-types';
+import {useService} from '$lib/services/service-provider';
+import type {IJsEventListener} from '$lib/dotnet-types/generated-types/FwLiteShared/Events/IJsEventListener';
+import type {IFwEvent} from '$lib/dotnet-types/generated-types/FwLiteShared/Events/IFwEvent';
+import {FwEventType} from '$lib/dotnet-types/generated-types/FwLiteShared/Events/FwEventType';
+import type {IEntryChangedEvent} from '$lib/dotnet-types/generated-types/FwLiteShared/Events/IEntryChangedEvent';
+import type {IProjectEvent} from '$lib/dotnet-types/generated-types/FwLiteShared/Events/IProjectEvent';
 
 export class EventBus {
-  private _onEntryUpdated = new Set<(entry: IEntry) => void>();
+  private _onEvent = new Set<(event: IFwEvent) => void>();
   private _onProjectClosed = new Set<(reason: CloseReason) => void>();
+  constructor() {
+    void this.eventLoop(useService(DotnetService.JsEventListener));
+  }
+
+  private async eventLoop(jsEventListener: IJsEventListener) {
+    let event: IFwEvent;
+    do {
+      event = await jsEventListener.nextEventAsync();
+      if (!event) return;
+      this.distributor(event);
+    } while (event);
```

Comment:
I find this a bit clearer
```suggestion
    while (true) {
      event = await jsEventListener.nextEventAsync();
      if (!event) return;
      this.distributor(event);
    }
```